### PR TITLE
Megaphones

### LINKED ISSUE #2362: Megaphones
# Description

Megaphones let the crewmembers know who's **really** in charge of the station, by making a users text appear bolded and larger than most voices (font-size: 3). They only apply while they are in the active hand, and also apply through telecomms as well. On use, they play a sound effect and set a small timeout before it can be used again. It can be emagged to be REALLY BIG (font-size: 4).

Along with the normal megaphone, there are a few variants:
- Security
- Command
- Cargo
- Clown

The clown megaphone makes the users text big AND comic sans AND bright pink. (color: #FF69Bf;	font-size: 3; font-family: "Comic Sans MS", cursive, sans-serif; font-weight: bold;)

https://github.com/tgstation/tgstation/blob/5481515dff73d67f1d1a60b4a6fd3883d034c456/code/game/objects/items/devices/megaphone.dm

### PR DESCRIPTION
**Pull request by MaxIsJoe**
CL: [New] Adds in Megaphones! with three variants for departments.
Improvements for the chat system to handle items that are chat related better and made creating chat based items much easier.

### Notes :
Add the clown variant later with it's own functionality
Add sounds later for all megaphones

closes #2362

### CODE REVIEW

**Comment by MaxIsJoe** [comment]
We're ready for review!

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs [hunk 1]
```diff
@@ -232,6 +233,20 @@ public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message
 			}
 		}
 
+		//Check if there's any items on the player that affects their chat (e.g : headphones, muzzles, etc)
+		foreach (var slots in player.DynamicItemStorage.ServerContents.Values)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: might error if you're a ghost?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Fixed it

**Comment by MaxIsJoe** [comment]
approvals pls

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs [hunk 2]
```diff
@@ -233,16 +233,19 @@ public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message
...
+			foreach (var slots in player.DynamicItemStorage.ServerContents.Values)
 			{
-				if (slot.IsEmpty) continue;
-				if (slot.Item.TryGetComponent<IChatInfluncer>(out var listener)
-				    && listener.RunChecks() == true)
+				foreach (var slot in slots)
 				{
-					chatEvent = listener.InfluenceChat(chatEvent);
+					if (slot.IsEmpty) continue;
+					if (slot.Item.TryGetComponent<IChatInfluncer>(out var listener)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: What is a "chat influncer"?

**Review by NoooneyDude — DISMISSED** [review]

File: UnityProject/Assets/Scripts/Core/Chat/Chat.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: A thing that influences the chat going through

So for megaphone's example, when the `InfluenceChat()` function gets called it influences something in the ChatEvent that is being sent. In this case we are influencing the `voiceLevel` of the chat event.

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: It's also an interface not an abstract class

Comment (corp-0) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: So a chat influ**e**ncer then, no?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Core/Chat/Chat.cs]: Ah, fixed the type. thanks for pointing that out.

**Review by NoooneyDude — DISMISSED** [review]

**Review by NoooneyDude — APPROVED** [review]

**Pull request merged**